malcolm/core/process.py

```diff
@@ -91,3 +109,84 @@ def spawn(self, function, *args, **kwargs):
         self._other_spawned.append(spawned)
         return spawned
 
+    def _handle_block_notify(self, request):
+        """Update subscribers with changes and applies stored changes to the
+        cached structure"""
+        # update cached dict
+        for path, value in self._last_changes.setdefault(request.name, []):
+            d = self._block_state_cache
+            for p in path[:-1]:
+                d = d[p]
+            d[path[-1]] = value
+
+        for subscription in self._subscriptions.setdefault(request.name, []):
+            endpoint = subscription.endpoint
+            # find stuff that's changed that is relevant to this subscriber
+            changes = []
+            for change_path, change_value in self._last_changes[request.name]:
+                # look for a change_path where the beginning matches the
+                # endpoint path, then strip away the matching part and add
+                # to the change set
+                i = 0
+                for (cp_element, ep_element) in zip(change_path, endpoint):
+                    if cp_element != ep_element:
+                        break
+                    i += 1
+                else:
+                    # change has matching path, so keep it
+                    # but strip off the end point path
+                    filtered_change = [change_path[i:], change_value]
+                    changes.append(filtered_change)
+            if len(changes) > 0:
+                if subscription.delta:
+                    # respond with the filtered changes
+                    response = Response.Delta(
+                        subscription.id_, subscription.context, changes)
+                    subscription.response_queue.put(response)
+                else:
+                    # respond with the structure of everything
+                    # below the endpoint
+                    update = self._block_state_cache
+                    for p in endpoint:
+                        update = update[p]
+                    response = Response.Update(
+                        subscription.id_, subscription.context, update)
+                    subscription.response_queue.put(response)
+        self._last_changes[request.name] = []
+
+    def _handle_block_changed(self, request):
+        """Record changes to made to a block"""
+        for path, value in request.changes:
```

Comment:
Consider:
[["a"],{"b": 32}], [["a", "b"], 42], [["a"], {"b": 52}]

In your code this is compressed to: 
[["a"],{"b": 52}], [["a", "b"], 42]

Which gives a different result.

I think we shouldn't bother with squashing duplicate updates, just pass them through to the client...